lib/source/consul.js

```diff
@@ -0,0 +1,292 @@
+      });
+
+      watcher.on('change', (healthServiceData) => {
+        this._onHealthServiceChange(info.name, healthServiceData);
+      });
+
+      this._registerHealthWatcher(info.name, watcher);
+    });
+
+    this._okay = true;
+  }
+
+  /**
+   * Process updates from Consul's health/service API.
+   *
+   * The data returned from Consul has the following format:
+   * [{
+   *   "Node": {
+   *     "Address": "127.0.0.1"
+   *   },
+   *   "Service": {
+   *     "Address": ""
+   *   }
+   * }]
+   *
+   * The Service.Address attribute is provided may or may not be given.
+   * If it's provided, it's the address defined by the service. The
+   * Node.Address attribute is the address provided by Consul.
+   *
+   * @param {String} name  The name of the watcher
+   * @param {Object} data  The JSON data from the health/service API
+   * @private
+   */
+  _onHealthServiceChange(name, data) {
+    if (data.length > 0) {
+      const addresses = [];
+
+      for (let i = 0; i < data.length; i += 1) {
+        const info = data[i];
+
+        if (info.Service && info.Service.Address) {
+          addresses.push(info.Service.Address);
+        } else if (info.Node && info.Node.Address) {
+          addresses.push(info.Node.Address);
+        }
+      }
+
+      if (!this.properties[name]) {
+        this.properties[name] = Object.create(null);
+      }
+      this.properties[name].addresses = addresses.sort();
+    } else {
+      // Empty data means the service has been deregistered.
+      this._unregisterHealthWatcher(name);
+      delete this.properties[name];
+    }
+
+    this._okay = true;
+    this.emit('update', this.properties);
+  }
+
+  /**
+   * Checks if there's alread a watcher for the Consul health/service API
+   *
+   * @param {String}    name The name of the watcher to check
+   * @return {Boolean}  true if there's a watcher with the name; false otherwise
+   * @private
+   */
+  _hasHealthWatcher(name) {
+    return this._healthWatchers && this._healthWatchers[name];
+  }
+
+  /**
+   * Register a watcher that's tracking a Consul health/service API
+   * @param {String} name           The name of the watcher
+   * @param {Consul.Watch} watcher  The watcher returned by `consul.watch()`
+   * @private
+   */
+  _registerHealthWatcher(name, watcher) {
+    if (!this._healthWatchers) {
+      this._healthWatchers = Object.create(null);
+    }
+    this._healthWatchers[name] = watcher;
+  }
+
+  /**
+   * Unregister a watcher that's tracking a Consul health/service API
+   * @param {String} name  The name of the watcher
+   * @private
+   */
+  _unregisterHealthWatcher(name) {
+    if (this._hasHealthWatcher(name)) {
+      this._healthWatchers[name].end();
+      delete this._healthWatchers[name];
+    }
+  }
+
+  /**
+   * Shut down everything tracking the Consul health/service API.
+   * @private
+   */
+  _shutdownHealthWatchers() {
+    for (const name in this._healthWatchers) {
+      if (this._healthWatchers[name]) {
```

Comment:
I don't think you need this test if you're already iterating over `this._healthWatchers`, especially since calling `_unregisterHealthWatcher()` has the same check.